### PR TITLE
Wasm Light Client

### PR DESCRIPTION
**Pull request by jackzampolin**
Major areas for work:
- [x] Clean up existing implementation abstractions. This is mainly around how the light client code interacts with the 28-wasm vm instance. Currently there are very leaky abstractions and a usage of a global package scope variable which is likely to cause problems. 
- [x] Define the interface that is expected of the contract by the 28-wasm client. Will add comments here to describe where the calls are what they look like. This maybe should be modified to fit updated cosmwasm development idioms.
- [x] Add governance proposal to enable the chain to add new wasm clients to use
- [ ] build out unit/integration test
- [x] merge in typed interface
- [x] re-add status method to get status from contract
- [x] adjust wasmVM parameters and gas consumption

Unit testing:
- [x] Status
- [x] Validate
- [x] Initialize
- [x] VerifyMembership
- [x] VerifyNonMembership
- [x] ExportMetadata
- [x] VerifyHeader
- [x] UpdateState
- [ ] PruneConsensusState
- [x] CheckForMisbehaviour
- [x] UpdateStateOnMisbehaviour
- [x] VerifyUpgrade (only testing contract interface)
- [x] VerifyMisbehaviour
- [x] MisbehaviourValidateBasic
- [x] CheckSubstituteUpdateStateBasic
- [x] CheckSubstituteAndUpdateState
- [x] ConsensusStateValidateBasic
- [x] HeaderValidateBasic

### CODE REVIEW

File: modules/light-clients/10-wasm/types/client_state.go [hunk 1]
```diff
@@ -226,57 +203,12 @@ func (c *ClientState) UpdateStateOnMisbehaviour(ctx sdk.Context, cdc codec.Binar
...
-// 	if !output.Result.IsValid {
-// 		return nil, fmt.Errorf("%s error occurred while updating client state", output.Result.ErrorMsg)
-// 	}
-// 	output.resetImmutables(c)
-// 	return output.NewClientState, nil
-// }
-// CheckSubstituteAndUpdateState(ctx sdk.Context, cdc codec.BinaryCodec, subjectClientStore, substituteClientStore sdk.KVStore, substituteClient ClientState) error
-
 func (c *ClientState) UpdateState(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, clientMsg exported.ClientMessage) []exported.Height {
+	_, err := SaveClientStateIntoWasmStorage(ctx, cdc, clientStore, c)
```
Comment (blasrodri) [inline modules/light-clients/10-wasm/types/client_state.go:207]: I guess this assumes that the light client is stateless? Otherwise, if the LC itself is running the update state logic and updating the storage itself, this call won't be needed right?

Comment (jackzampolin) [reply modules/light-clients/10-wasm/types/client_state.go:207]: @blasrodri we have yet to fully define the interface so this code can be considered WIP but the current implementation should be duplication of storage

File: modules/light-clients/10-wasm/types/client_state.go [hunk 2]
```diff
@@ -194,21 +189,68 @@ func (c *ClientState) VerifyClientMessage(ctx sdk.Context, cdc codec.BinaryCodec
...
+	}
+
+	// Get consensus bytes from clientStore
+	consensusState1, err := GetConsensusState(clientStore, cdc, wasmMisbehaviour.Header1.Height)
+	if err != nil {
+		return false
+	}
+
+	// Get consensus bytes from clientStore
+	consensusState2, err := GetConsensusState(clientStore, cdc, wasmMisbehaviour.Header2.Height)
```
Comment (vmarkushin) [inline modules/light-clients/10-wasm/types/client_state.go:204]: Why do you think we need consensus states to check misbehavior?

File: modules/light-clients/10-wasm/types/client_state.go [hunk 3]
```diff
@@ -194,21 +189,68 @@ func (c *ClientState) VerifyClientMessage(ctx sdk.Context, cdc codec.BinaryCodec
...
+	payload[checkForMisbehaviourMessage] = make(map[string]interface{})
+	inner := payload[checkForMisbehaviourMessage]
+	inner["me"] = c
+	inner["misbehaviour"] = wasmMisbehaviour
+	inner["consensus_state1"] = consensusState1
+	inner["consensus_state2"] = consensusState2
+
+	encodedData, err := json.Marshal(payload)
+	if err != nil {
+		return false
```
Comment (vmarkushin) [inline modules/light-clients/10-wasm/types/client_state.go:220]: Shoudn't we panic in this case?

File: modules/light-clients/10-wasm/types/client_state.go [hunk 4]
```diff
@@ -194,21 +189,68 @@ func (c *ClientState) VerifyClientMessage(ctx sdk.Context, cdc codec.BinaryCodec
...
+	payload[updateStateOnMisbehaviour] = make(map[string]interface{})
+	inner := payload[updateStateOnMisbehaviour]
+	inner["client_state"] = c
+	inner["client_message"] = clientMsg
+
+	encodedData, err := json.Marshal(payload)
+	if err != nil {
+		return
+	}
+	callContract(c.CodeId, ctx, clientStore, encodedData)
```
Comment (vmarkushin) [inline modules/light-clients/10-wasm/types/client_state.go:249]: Result handling?

Comment (antstalepresh) [reply modules/light-clients/10-wasm/types/client_state.go]: added panic for contract call failure

File: modules/light-clients/10-wasm/types/client_state.go (hunk 2 above)
Comment (blasrodri) [reply modules/light-clients/10-wasm/types/client_state.go:204]: It's used here for example https://github.com/strangelove-ventures/ibc-go/blob/f8b4828f3b2b909edcef22ad31b153cfc4d98b42/modules/light-clients/07-tendermint/update.go#L208

Comment (vmarkushin) [reply modules/light-clients/10-wasm/types/client_state.go]: I think it's more implementation-specific and a client can read the consensus state from the storage on demand

Comment (blasrodri) [reply modules/light-clients/10-wasm/types/client_state.go]: I agree with that

**Comment by blasrodri** [comment]
After the walkthrough that @misko9 gave to the ibc-go team there were some outstanding questions that we need to address:

1. what’s the reason for using the wasmvm directly instead of wasmd? Particularly, given that [wasmd has the functionality to push code, which this PR is implementing](https://github.com/CosmWasm/wasmd/blob/c7670fae4e6ca24d7fd6bfa6768f166b36976432/x/wasm/keeper/msg_server.go#L22).
2. impact in terms of performance of having to run 2 vms in parallel for CW-enabled chains?

**Comment by blasrodri** [comment]
I've added a benchmark comparing execution of a large amount of calls to the wasm code using ONE VM vs TWO VMs
https://github.com/blasrodri/libwasm-benchmark

```
RUST_BACKTRACE=1 go test -v -bench=.
goos: linux
goarch: amd64
pkg: github.com/CosmWasm/wasmvm
cpu: AMD Ryzen 9 5900X 12-Core Processor            
BenchmarkHappyPathOneVM
BenchmarkHappyPathOneVM-24             1        2209188026 ns/op
BenchmarkHappyPathTwoVMs
BenchmarkHappyPathTwoVMs-24            1        2295065363 ns/op
PASS
ok      github.com/CosmWasm/wasmvm      4.513s
```
Seems to be negligible

**Comment by MuhammedIrfan** [comment]
Hi Team, We are very excited to see this PR, This will help bring more chains to IBC ecosystem. Do you any planned timeframe that we can expect this PR to merged?

**Comment by faddat** [comment]
We're grinding till it's done @MuhammedIrfan

**Comment by faddat** [comment]
@blasrodri I think that the grind will just move to another core.  We will get some of that cgo overhead, but likely nbd.

**Comment by faddat** [comment]
Hi @misko9 -- this looks good!  Nice work on the db converter :)

**Comment by faddat** [comment]
this is beautiful

**Comment by misko9** [comment]
@faddat 
I don't know for sure, but my guess is yes, it's v8. The upstream PR is tagged with v7.2.0, but target branch is main. I expect it will also merge into a newly created release/v7.2.x branch once approved. I'll be maintaining both branches (main and release/v7.1.x baselines) in Strangelove's repo.

File: modules/light-clients/08-wasm/types/vm.go [hunk 5]
```diff
@@ -91,14 +91,14 @@ func callContract(codeID []byte, ctx sdk.Context, store sdk.KVStore, msg []byte)
 	if height < 0 {
 		panic("Block height must never be negative")
 	}
-	sec := ctx.BlockTime().Unix()
-	if sec < 0 {
```
Comment (blasrodri) [inline modules/light-clients/08-wasm/types/vm.go]: Why the change?

Comment (misko9) [reply modules/light-clients/08-wasm/types/vm.go]: Discovered the wasm vm expects time in nanos. Found while testing, as I go, the tendermint cosmwasm contract I've been working on. It has dependencies on the correct time being passed in for the delay time period and the block time at which a consensus state/header is processed.

**Comment by akc2267** [comment]
closing as this is merged into a feature branch already
